### PR TITLE
Improve compatibility between preprocessing defences and standardisation

### LINKED ISSUE #84: Compatibility of defence JpegCompression with classifier's preprocessing (standardisation)
**Is your feature request related to a problem? Please describe.**
- In classifer.predict we are applying the classifier's preprocessing before applying the defences.
- Most of the common classifier preprocessing values for standardisation result in pixel values outside of [0, 1].
- The defence `JpegCompression` expects inputs `x` to be in the range [0, 1] and clips inputs `x` to this range [0, 1].
- This can very likely result in unexpected behaviour of the classifier without any warnings.

**Describe the solution you'd like**
Are there any reasons against reversing the order to first applying the defences and then applying the preprocessing (standardisation) in classifier.predict()?
For example are there other defences that need to run after preprocessing (standardisation)?

**Describe alternatives you've considered**
- As alternative we could use the classifier's clip_values in JpegCompression to adapt the clipping for the compression step instead of just assuming [0, 1] for `x`.
- We could reverse the standardisation in `JpegCompression` and reapply it after the compression step, but this would result in extra computing efforts.
- At least we should print a warning if `JpegCompression` clips values in the input `x` that are outside of the expected range.

**Additional context**
The behaviour of `JpegCompression` is documented in its docstring.

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This PR changes the order of the preprocessing steps of applying defences and standardisation/normalisation. Up to now the classifiers first applied standardisation followed by defences. This can result in unexpected negative and/or unbounded input values for the defences. With this PR the defences will be applied first followed by standardisation, that way we can feed to unprocessed input image values to the classifier and use the same and comparable defence parameters across classifiers with different standardisation/normalisation parameters.

Fixes #84 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [ ] Test A
- [ ] Test B

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by beat-buesser** [comment]
I need to fix some unit tests, see #95, before the reviews can start.

**Comment by beat-buesser** [comment]
Fixes #95

**Comment by minhitbk** [comment]
Changes look fine to me, its ready for merging.

**Comment by beat-buesser** [comment]
@ririnicolae Thank you for your initial feedback.

The change in the design of the defences originated in the idea to treat all attributes in the same way.

I don't think this puts an additional burden on the users because they will still be able to use the default values for the corresponding arguments and don't need to configure them.

I agree that that all variables should now be public, to remove the property for `apply_fit` and `apply_predict`, and to make it consistent among all defences. I'll push an update soon.
